src/replication.c

```diff
@@ -750,6 +766,225 @@ is_replica_newly_connected(spec_t *spec, replica_t *new_replica)
+	tmp_data = (char *)malloc(sizeof(char) * 64);
+	rc = istgt_readline_socket(fd, read_data, 128, tmp_data, 64, &tmpidx, &tmpcnt, rtimeout);
+	if (rc < 0) {
+		ISTGT_ERRLOG("failed to read the data on %d\n", fd);
+		goto clean_data;
+	}
+	if(strcmp(read_data, err) == 0 || strlen(read_data) == 0) {
+		ISTGT_ERRLOG("failed to update %s\n", read_data);
+		rc = -1;
+		goto clean_data;
+	}
+	rc = 1;
+clean_data:
+	free(read_data);
+	free(tmp_data);
+error_in_update:
+	close(fd);
+	return rc;
+#endif
+}
+
+// update_in_memory_known_replica_list will update replica zvol guid
+// if replicaId is already exist else it will add new known replica to list
+static int
+update_in_memory_known_replica_list(spec_t *spec, replica_t *replica) {
+	trusty_replica_t *knr;
+	int key_len;
+
+	ASSERT(MTX_LOCKED(&spec->rq_mtx));
+
+	TAILQ_FOREACH(knr, &spec->lu->known_replica, next) {
+		if (strcmp(knr->replica_id, replica->replica_id) == 0) {
+			knr->zvol_guid = replica->zvol_guid;
+			return 0;
+		}
+	}
+
+	knr = xmalloc(sizeof (trusty_replica_t));
+	memset(knr, 0, sizeof(trusty_replica_t));
+	key_len = strlen(replica->replica_id);
+	knr->replica_id = (char *)malloc(sizeof(char) * key_len);
+	if (knr->replica_id == NULL) {
+		ISTGT_ERRLOG("failed to allocate memory for known"
+		    " replicaId %s\n", replica->replica_id);
+		return -1;
+	}
+	strncpy(knr->replica_id, replica->replica_id, key_len);
+	knr->zvol_guid = replica->zvol_guid;
+	TAILQ_INSERT_TAIL(&spec->lu->known_replica, knr, next);
+	return 0;
+}
+
+/* update_known_replica_list will be called only once in replicas
+ * life time (even it is disconnected or connected back n times)
+ */
+static int
+update_known_replica_list(spec_t *spec, replica_t *rep) {
+	int replica_count, data_len, rc;
+	char *data;
+	const char *json_string;
+	struct json_object *jobj;
+
+	replica_count = spec->healthy_rcount + spec->degraded_rcount;
+
+	if (replica_count >= spec->replication_factor) {
+		ISTGT_ERRLOG("known replica count %d is greater "
+		    "than or equal to replciation factor %d\n",
+		    replica_count, spec->replication_factor);
+		return -1;
+	}
+
+	jobj = json_object_new_object();
+	json_object_object_add(jobj, "replicaId",
+	    json_object_new_string(rep->replica_id));
+	json_object_object_add(jobj, "replicaZvolGuid",
+	    json_object_new_uint64(rep->zvol_guid));
+	json_object_object_add(jobj, "volumeName",
+	    json_object_new_string(spec->volname));
+	json_string = json_object_to_json_string_ext(jobj, JSON_C_TO_STRING_PLAIN);
+
+	data_len = strlen(json_string) + 1;
+	data = (char *) malloc(sizeof(char *)*data_len);
+	memset(data, 0, data_len);
+	strncpy(data, json_string, data_len);
+
+	rc = update_volume_config(data);
+	if (rc < 0) {
+		ISTGT_ERRLOG("failed to update replica(%s:%lu) to known replica list\n",
+		    rep->replica_id, rep->zvol_guid);
+	} else {
+		(void) update_in_memory_known_replica_list(spec, rep);
+	}
+
+	free(data);
+	json_object_put(jobj);
```

Comment:
do we need to free json_string also? cc: @utkarshmani1997 can you please look at json related things wrt any pointer corruption/mem leak?

Reply:
@vishnuitta looks good, freeing root object will free all the allocated memory associated with the json_object.